cnd/src/swap_protocols/herc20.rs

```diff
@@ -0,0 +1,263 @@
+use crate::{
+    asset,
+    swap_protocols::{
+        rfc003::{Secret, SecretHash},
+        state, SwapId,
+    },
+};
+use blockchain_contracts::ethereum::rfc003::Erc20Htlc;
+use chrono::NaiveDateTime;
+use futures::{
+    future::{self, Either},
+    Stream,
+};
+use genawaiter::sync::{Co, Gen};
+use std::collections::{hash_map::Entry, HashMap};
+use tokio::sync::Mutex;
+
+mod connector;
+
+pub use connector::*;
+
+/// Resolves when said event has occured.
+#[async_trait::async_trait]
+pub trait Deployed {
+    async fn deployed(&self, params: Params) -> anyhow::Result<data::Deployed>;
+}
+
+#[async_trait::async_trait]
+pub trait Funded {
+    async fn funded(
+        &self,
+        params: Params,
+        deployed: data::Deployed,
+    ) -> anyhow::Result<data::Funded>;
+}
+
+#[async_trait::async_trait]
+pub trait Redeemed {
+    async fn redeemed(
+        &self,
+        params: Params,
+        deployed: data::Deployed,
+    ) -> anyhow::Result<data::Redeemed>;
+}
+
+#[async_trait::async_trait]
+pub trait Refunded {
+    async fn refunded(
+        &self,
+        params: Params,
+        deployed: data::Deployed,
+    ) -> anyhow::Result<data::Refunded>;
+}
+
+/// Represents states that an ERC20 HTLC can be in.
+#[derive(Debug, Clone)]
+pub enum State {
+    None,
+    Deployed(data::Deployed),
+    Funded(data::Funded),
+    Redeemed(data::Redeemed),
+    Refunded(data::Refunded),
+}
+
+/// Represents the events in the herc20 protocol.
+#[derive(Debug, Clone, PartialEq, strum_macros::Display)]
+pub enum Event {
+    /// The protocol was started.
+    Started,
+
+    /// The HTLC was deployed and is pending funding.
+    Deployed(data::Deployed),
+
+    /// The HTLC has been funded with ERC20 tokens.
+    Funded(data::Funded),
+
+    /// The HTLC has been destroyed via the redeem path, token have been sent to
+    /// the redeemer.
+    Redeemed(data::Redeemed),
+
+    /// The HTLC has been destroyed via the refund path, token has been sent
+    /// back to funder.
+    Refunded(data::Refunded),
+}
+
+/// Represents the data available at said state.
+pub mod data {
```

Comment:
The `data` module is bugging me for `halight` as-well. Can we get rid of this?

The naming conflict should urge us to think about a better name for one of the two concepts instead of moving one of the types into a module with a generic name.

To start with, I would say the names of the events are more naturally expressed as they currently are so I would probably change the names of the traits. What about `WaitForFunded` etc?
If we find a way for either not passing in `Deployed` separately, we could even to something like `WaitFor<Funded>` :exploding_head:

Reply:
> If we find a way for either not passing in `Deployed` separately, we could even to something like `WaitFor<Funded>` 🤯

There are ways but it gives less type safety so for now I'll leave it with `WaitFor[..]` traits